### DISCUSSION TITLE
Proposal to Create Canvas-only bundle

### DISCUSSION THREAD

**Discussion by bigtimebuddy** [pixijs/pixijs#8477] [RFC]
## Overview

We think there's some utility in having a CanvasRenderer only bundle (think `pixi.js-legacy` but without WebGL). This bundle would contain only the canvas supported bits (Texture, BaseTexture, RenderTexture, etc), but none of the WebGL specifics (filters, Multisample, Buffers).

The benefit of doing this would mean that we'd have a _very_ small version of PixiJS. Most of the WebGL code takes up the majority of the footprint. If someone needed a maximally small version of PixiJS without all the WebGL bells and whistles, then this is your solution! This would also work well in certain environments (like embedded hardware such as Raspberry PI, where getting GL to run is difficult) or potentially a less memory hungry Node.js implementation.

This is architecturally a difficult task and involves moving a lot of pieces around, so we want to know if folks would find this feature useful. **Would you use this?** If so, could you describe your use-cases.

The side benefit of doing this work would also mean that introducing a WebGPU renderer would be easier as WebGL is not considered the default core renderer.

## Details

* We need to split `@pixi/core` into two pieces: `@pixi/core` and `@pixi/webgl-renderer`
* The `@pixi/core` would largely consist of these pieces (not an exhaustive list):
  * SystemManager
  * MaskData
  * IRenderer
  * RenderTexture
  * IRenderableObject
  * BackgroundSystem
  * ViewSystem
  * BaseTexture
  * BaseRenderTexture
  * RenderTexture
  * Texture
  * extensions
  * ExtensionType
  * ExtensionMetadata
  * CanvasResource
  * ImageBitmapResource
  * ImageResource
  * SVGResource
  * VideoResource
  * ISystem
  * StartupSystem
  * StartupOptions
  * PluginSystem
  * IGenerateTextureOptions
  * GenerateTextureSystem
* Create a new Canvas-only bundle `@pixi/canvas` or `pixi.js-canvas`, would have most of the features of `pixi.js-legacy` with the exception of:
  * `@pixi/compressed-textures`
  * `@pixi/filter-*`
  * `@pixi/extract`
  * `@pixi/prepare`
* Need to mixin the WebGL-specific `render()` functions into display objects (Sprite, Mesh, Graphics, etc) or find a generic way to describe this using IRenderer interface.

**Comment by luxaritas** [comment]
How does this proposal interact with prior comments about intending to removing canvas entirely to reduce complexity and maintenance burden? How does this not run into the same issues? Or does it, with the proposal being to reverse course due to the potential benefits? Is this a matter of not having fully-compatible WebGL and Canvas implementations, so newer features don't need to be "backported"?

**Reply by bigtimebuddy** [reply]
> How does this proposal interact with prior comments about intending to removing canvas entirely to reduce complexity and maintenance burden? 

We have reversed course on the canvas issue. There are still appropriate uses and some environments where it's still helpful. Canvas also has antialiasing for free, which is a pain in WebGL. We have found other ways to reduce the maintenance cost (see next response).

> How does this not run into the same issues? Or does it, with the proposal being to reverse course due to the potential benefits? 

We have already refactored v7 to make Renderers more compose-able. Before CanvasRenderer had a lot of duplication of code with WebGL Renderer, but that's not the case anymore. And we have a plan to add WebGPU in a way that's manageable.

> Is this a matter of not having fully-compatible WebGL and Canvas implementations, so newer features don't need to be "backported"?

Canvas will always be a subset of WebGL features because it's less flexible. The advantage is in the overall size. It's potentially super small on its own.

**Reply by luxaritas** [reply]
Makes sense! Thanks for taking the time to explain. 🙂

**Comment by goldenratio** [comment]
I doubt we would need this for game development. Maybe for eye-candy stuff for websites.

**Comment by ivanpopelyshev** [comment]
Consider our current level of documentation (high) , and things that have to be done for big refactoring to not lose our quality, its better to spend this time on actual webgl/webgpu hybrid than further decoupling stage from middle from low level

WebGL and WebGPU actually are possible to make into one interface. We have to add state approach to webgpu and immutable approach to webgl.
